### PR TITLE
Increase performance of OneToOneRingBuffer...

### PR DESCRIPTION
**Pull request by jfallows**
… by eliminating setMemory during read.

The call to setMemory was used by the reader to zero the bytes of all read messages so that subsequently written variable length messages are guaranteed to be followed by a zero length record header. The reader observes this zero length record header to efficiently detect when to stop reading.

The responsibility for each message to be followed by a zero length record header is now shifted to the writer, which already knows the precise location of the end of each message record.

The reader now no longer writes a block of zeros to the record buffer during read.

The writer requires each message to be followed by a zero length record header, even when the ring buffer is full, so the total capacity is reduced by the size of a record header (8 bytes).

### CODE REVIEW

**Comment by nitsanw** [comment]
See the discussion here:
https://github.com/real-logic/Agrona/issues/34

**Comment by jfallows** [comment]
@nitsanw: thanks.

**Comment by jfallows** [comment]
@nitsanw: I understand from #34 that the MTORB reader calls `setMemory` to ensure that a zero length header implicitly follows new variable-length messages written to the ring buffer, even though the exact end of those messages is not predictable (since they are variable length).

> The Zero'ing is to ensure the message length will be zero for variable length messages.

This patch to OTORB maintains the same invariant without requiring the reader to call `setMemory`.

File: src/test/java/org/agrona/concurrent/ringbuffer/OneToOneRingBufferTest.java [hunk 1]
```diff
@@ -186,8 +194,8 @@ public void shouldNotReadSingleMessagePartWayThroughWriting()
 
         final InOrder inOrder = inOrder(buffer);
         inOrder.verify(buffer, times(1)).getLongVolatile(headIndex);
-        inOrder.verify(buffer, times(0)).setMemory(headIndex, 0, (byte)0);
         inOrder.verify(buffer, times(0)).putLongOrdered(HEAD_COUNTER_INDEX, headIndex);
+        inOrder.verify(buffer, times(0)).setMemory(anyInt(), anyInt(), anyByte());
```
Comment (nitsanw) [inline src/test/java/org/agrona/concurrent/ringbuffer/OneToOneRingBufferTest.java]: pointless reordering of lines here?

File: src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java [hunk 2]
```diff
@@ -85,7 +85,8 @@ public boolean write(final int msgTypeId, final DirectBuffer srcBuffer, final in
 
         final AtomicBuffer buffer = this.buffer;
         final int recordLength = length + HEADER_LENGTH;
-        final int requiredCapacity = align(recordLength, ALIGNMENT);
+        final int alignedRecordLength = align(recordLength, ALIGNMENT);
+        final int requiredCapacity = alignedRecordLength + HEADER_LENGTH;
```
Comment (nitsanw) [inline src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java]: Why do we require the extra capacity?
Before: requiredCapacity = align(length + HEADER_LENGTH, ALIGNMENT)
After:    requiredCapacity = align(length + HEADER_LENGTH, ALIGNMENT) + HEADER_LENGTH

File: src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java [hunk 3]
```diff
@@ -132,13 +133,15 @@ public boolean write(final int msgTypeId, final DirectBuffer srcBuffer, final in
 
         if (0 != padding)
         {
+            buffer.putLong(0, 0L);
```
Comment (nitsanw) [inline src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java]: What are you zeroing at index 0?

File: src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java [hunk 4]
```diff
@@ -132,13 +133,15 @@ public boolean write(final int msgTypeId, final DirectBuffer srcBuffer, final in
...
             buffer.putLongOrdered(recordIndex, makeHeader(padding, PADDING_MSG_TYPE_ID));
             recordIndex = 0;
         }
 
         buffer.putBytes(encodedMsgOffset(recordIndex), srcBuffer, srcIndex, length);
+        buffer.putLong(recordIndex + alignedRecordLength, 0L);
```
Comment (nitsanw) [inline src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java]: This is a nice solution, zeroing the next message header. This, combined with the extra capacity requirement should work, but at the cost of reducing buffer actual capacity by 8 bytes. You can add a condition or accept the loss.

File: src/test/java/org/agrona/concurrent/ringbuffer/OneToOneRingBufferTest.java [hunk 5]
```diff
@@ -301,7 +309,7 @@ public void shouldThrowExceptionForCapacityThatIsNotPowerOfTwo()
     {
         final int capacity = 777;
         final int totalBufferLength = capacity + RingBufferDescriptor.TRAILER_LENGTH;
-        new ManyToOneRingBuffer(new UnsafeBuffer(new byte[totalBufferLength]));
+        new OneToOneRingBuffer(new UnsafeBuffer(new byte[totalBufferLength]));
```
Comment (nitsanw) [inline src/test/java/org/agrona/concurrent/ringbuffer/OneToOneRingBufferTest.java]: This is an unrelated fix.

**Comment by nitsanw** [comment]
LGTM with the following notes/exceptions:

  1. Need to justify the zeroing of index 0 when padding. I think the motivation is the zero the post pad header, but that's where the new message header is going to be so if that's the case it's redundant.
  2. In future, a short description of how your change works might help review.
  2. The actual capacity of the buffer is reduced by 8 bytes. Not a big deal surely, but worth mentioning as a side effect.
  3. I know @mjpt777 has very minimalist documentation preferences, but I personally could have done with some hints (e.g. // zero next message header)

File: src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java (hunk 2 above)
Comment (jfallows) [inline src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java]: As you indicated later, this is to guarantee there is enough space to write a zero for the next message header, which will act as a terminator for the reader.

File: src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java (hunk 4 above)
Comment (jfallows) [inline src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java]: Thanks @nitsanw.
I'm okay with either approach, let me know if you need a change to the code.

File: src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java (hunk 3 above)
Comment (jfallows) [inline src/main/java/org/agrona/concurrent/ringbuffer/OneToOneRingBuffer.java]: This is subtle and as you say later, initially seems redundant.

This line is required to maintain the invariant that the reader observes a subsequent zero record header after each record with length greater than zero, otherwise it's a race.

When the reader observes the non-zero padding record length written at line 137 below, then this zeroing of index 0 at line 136 guarantees that the reader also observes the subsequent zero record header.

Later, when the reader observes the non-zero message record length written at line 143 below, then the zeroing after the message record at line 142 guarantees that the reader also observes the subsequent zero record header.

If line 136 is omitted, then the reader may observe the non-zero padding length without a subsequent zero record header before the message record and the subsequent zero record header are observable. This would break the invariant for the padding record.

**Comment by jfallows** [comment]
@nitsanw, I've added a high-level description of the changes (including a mention of the capacity reduction by 8 bytes), applied the feedback on the test code and also provided a more detailed response to your specific question about zeroing at record index 0.

Regarding the comments, I had initially planned to include something like you suggest, but then made a deliberate attempt to match the style of the code. Let me know if you would like further changes.

**Comment by nitsanw** [comment]
I leave it to @mjpt777 to merge, I'm not sure how this plays into current release timing.

**Comment by mjpt777** [comment]
Just getting round to reviewing. I was on holiday. For reference the original approach did ensure that a ManyToOneRingBuffer consumer could work on the same underlying buffer with either a OneToOneRingBuffer or ManyToOneRingBuffer producer(s). Does this still hold?

**Comment by tmontgomery** [comment]
@mjpt777 with this change I am unsure it would be possible. Unblocking behavior might be effected.

**Comment by mjpt777** [comment]
It should be possible to achieve these goals if the headers and counters are updated in the appropriate order.

BTW zeroing the next header is a nice touch.

**Comment by jfallows** [comment]
@mjpt777: a ManyToOneRingBuffer consumer still works with a OneToOneRingBuffer producer or with N different ManyToOneRingBuffer producer(s).

The OneToOneRingBuffer producer is writing the trailing zero header, and the ManyToOneRingBuffer consumer is also writing zeros over the data as it reads. The head position and tail position are being updated with the same semantics as before, so no change there.

@tmontgomery: the unblock behavior on the ManyToOneRingBuffer seems to have three distinct code paths; empty, negative length, and zero length.

The empty RingBuffer case is unaffected by this change.

The negative length is triggered by a ManyToOneRingBuffer producer, which has a two-phase claim-and-commit for writes. The OneToOneRingBuffer producer has a single-phase commit for writes, so no negative length case.

The zero length case has code that detects contiguous type id zero records to replace with a single padding type record. The contiguous zeros are still present, written by the ManyToOneRingBuffer consumer, so this code path also remains unaffected.

**Comment by tmontgomery** [comment]
@jfallows for OneToOne, there isn't a need for unblock. But when using the same buffer for ManyToOne, there could be. The state of the length field is only a very small part of the unblock behavior. The entire buffer state and the steps are quite important. So, adding a new step and removing the setMemory adds a new possibility of failure state(s) to (potentially) account for. So, will have to think about it to see if: (1) it matters, and (2) if it adds a new failure mode. The added `putLong` can be reordered, so it will float which adds a tiny bit of complexity.

**Comment by mjpt777** [comment]
To work with the many-to-one consumer, the producer will have to write the negative header length before the body. The consumer may unblock and get into a broken state. If this is addressed then it should work fine.

**Comment by jfallows** [comment]
Considering the scenario of OTORB producer with MTORB consumer, here are the relevant changes in this PR.
- OTORB capacity is reduced by 8 bytes
- OTORB writes a subsequent zero header (8 bytes) after each message
- OTORB subsequent zero header is observable with the same scope as record body bytes, i.e. before ordered record header and then ordered tail position updates

Notably, the MTORB consumer remains unchanged and therefore still calls setMemory to zero the bytes as it reads.

Considering the failure scenario where OTORB producer crashes just before writing the subsequent zero header, this should have unchanged behavior (no new code).

Considering the failure scenario where OTORB producer crashes just after writing the subsequent zero header, the record header with length and type id has not yet been written, therefore the record is not yet visible to the consumer and the tail position has not yet been updated either, so this should also have unchanged behavior.

Considering the failure scenario where OTORB producer crashes just after writing the record header with length and type id, the subsequent zero header is guaranteed to be observable by the consumer too, so this should also have unchanged behavior.

@mjpt777: if the MTORB consumer unblocks and gets into a broken state with a OTORB producer, is that due to the changes in this PR, or was that issue already present before these changes?

**Comment by mjpt777** [comment]
I see the tail index is updated at the end so the MTORB consumer would not try to unblock as it thinks no progress is being made. I think this if fine to merge. Thanks.
